### DISCUSSION TITLE
Where is state between DataReader::read() calls stored?

### DISCUSSION THREAD

**Discussion by ma30002000** [eProsima/Fast-DDS#4015] [Support]
Hi all,

I have a data reader with TopicQos and DataReaderQos configured DurabilityQosPolicyKind=TRANSIENT_LOCAL_DURABILITY_QOS and a history depth > 1. In general, accessing the history via the DataReader::read() method works quite well, however, when limiting the number of samples e.g. to 1 I always (and only) get the first sample stored in the data reader's history. 

I have tried calling DataReader::read() multiple times, with or without returning the loan, with the same LoanableCollection instance or with different ones for each call. However, I can still exclusively access the first sample in the history.

I was wondering 
a) what I might be doing wrong and 
b) where the state, which samples have already been read ist actually stored... In the LoanableCollections? I the data reader?

Thank you in advance for your help.

**Comment by MiguelCompany** [comment]
@ma30002000

> a) what I might be doing wrong

As you can read in our [documentation](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/subscriber/dataReader/readingData.html), there are two ways of accessing data: `read` and `take`.

The first one allows the user to peek into the head of the queue.
The second one pops data from the queue.

If the application needs to peek at unread data, setting the fourth argument of [`read`](https://fast-dds.docs.eprosima.com/en/latest/fastdds/api_reference/dds_pim/subscriber/datareader.html#_CPPv4N8eprosima7fastdds3dds10DataReader4readER18LoanableCollectionR13SampleInfoSeq7int32_t15SampleStateMask13ViewStateMask17InstanceStateMask) to `NOT_READ_SAMPLE_STATE` would do the trick.

> b) where the state, which samples have already been read is actually stored

This is stored internally in the DataReader. If you are curious and want to take a look at the code, it is stored in [this attribute](https://github.com/eProsima/Fast-DDS/blob/34c1b78a10d9d7b01ec5233d240afbe00cbd678a/include/fastdds/rtps/common/CacheChange.h#L90) of `CacheChange_t`. The DataReader keeps a collection of `CacheChange_t`, in its [`DataReaderHistory`](https://github.com/eProsima/Fast-DDS/blob/34c1b78a10d9d7b01ec5233d240afbe00cbd678a/src/cpp/fastdds/subscriber/history/DataReaderHistory.hpp#L59)

**Reply by ma30002000** [reply]
@MiguelCompany: Thank you for the fast response. So we read NOT_READ_SAMPLE_STATE samples making our way through the queue, with all already processed samples marked as "read" in the data reader.

What I want to do, however, is to read the history multiple times, i.e. the samples that are in READ_SAMPLE_STATE. If I understand your comment correctly it is only possible to read the "head of the queue", but not the whole queue? With the "head" being limited to the minimum between max_samples and ReaderResourceLimitsQos::max_samples_per_read and not possibility to access the samples that might rest in the queue after that? What would be the implication of settings this to larger values corresponding to the history (e.g. in the ranges 100-200 samples)?

**Reply by MiguelCompany** [reply]
Increasing `ReaderResourceLimitsQos::max_samples_per_read` will increase the memory usage of the DataReader. The other settings limiting the number of samples returned can have unbounded maximum values.

Other than the increase in memory usage, I see no other implications on increasing `max_samples_per_read`

**Comment by ma30002000** [comment]
Ok, thanks a lot!
